ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/blockimport/BlockImporter.java

```diff
@@ -37,32 +40,24 @@ public BlockImporter(ChainStorageClient storageClient, EventBus eventBus) {
     this.eventBus = eventBus;
   }
 
-  /**
-   * @param block The block to check
-   * @return {@code true} If this block is from the future according to our node's clock
-   */
-  public boolean isFutureBlock(BeaconBlock block) {
-    if (storageClient.isPreGenesis()) {
-      return true;
-    }
-    final UnsignedLong genesisTime = storageClient.getStore().getGenesisTime();
-    return ForkChoiceUtil.isFutureBlock(block, genesisTime, storageClient.getStore().getTime());
-  }
-
-  /**
-   * @param block The block to check.
-   * @return {@code true} If block is attached to the local chain
-   */
-  public boolean isBlockAttached(BeaconBlock block) {
-    return storageClient.getBlockState(block.getParent_root()).isPresent();
-  }
-
-  public void importBlock(BeaconBlock block) throws StateTransitionException {
+  public BlockImportResult importBlock(BeaconBlock block) {
     LOG.trace("Import block at slot {}: {}", block.getSlot(), block);
     Store.Transaction transaction = storageClient.startStoreTransaction();
-    final BlockProcessingRecord record = on_block(transaction, block, stateTransition);
-    transaction.commit().join();
-    eventBus.post(new BlockImportedEvent(block));
-    eventBus.post(record);
+    try {
+      final BlockProcessingRecord record = on_block(transaction, block, stateTransition);
+      transaction.commit().join();
+      eventBus.post(new BlockImportedEvent(block));
+      eventBus.post(record);
+    } catch (UnknownParentBlockException e) {
+      return BlockImportResult.create(e);
+    } catch (FutureBlockException e) {
+      return BlockImportResult.create(e);
+    } catch (InvalidBlockAncestryException e) {
+      return BlockImportResult.create(e);
+    } catch (StateTransitionException e) {
+      return BlockImportResult.create(e);
+    }
```

Comment:
It feels very weird to be throwing an exception and then having to convert to a return type.  Was it just that you didn't want to change the spec-derived code too much that it throws an exception or was switching to a return type all the way through too invasive?  If it's just that we didn't want to change the spec code, I'd be tempted to go change it to use return types.

To be clear, I think this is an improvement - it's basically an anti-corruption layer between spec code using exceptions and the rest of the world where we can use return types and live happily.

Reply:
Yeah, I guess I was trying to respect the existing style of the spec methods, but I did have the same thought after I pushed this up.  I'll try and push the `BlockImportResult` down to the `ForkChoicUtil` and see how that goes.

Reply:
Reworked it.